rfcs/records.md

```diff
@@ -0,0 +1,196 @@
+B1. Creation uses special syntax, `new Record(value, ...)`. This is unsurprising and easy to implement, but verbose.
+B2. Creation uses existing Lua call syntax, `Record(value, ...)`. This is concise and reasonably easy to keep efficient.
+
+The big difference between variants A and B is whether you need to spell out field names at construction time. There's precedents for going either way; some
+languages like F#/C#/Kotlin implement record construction as a function call and when you define a record, you essentially define the record constructor. This
+is beautifully concise, but is a bit more difficult to migrate away from tables, and it's easy to mix up the names. Variants A are more verbose and either require
+an extra `new` context-specific keyword, or complex/awkward magic to keep construction efficient.
+
+In variants A, it would make sense to allow omission of any field, in which case it gets replaced with the default of `nil`. A future extension could be made to
+allow specification of default values at record definition time. Type checker would fail to type check record construction if fields that have non-optional types
+have the values omitted.
+
+In variants B, it would probably make sense to require exact number of values to be specified, or follow the usual function call syntax rules.
+
+Note that since records are first class objects, you can export or import a record through a module boundary in the usual way:
+
+```
+local X = require(path).X
+local r = new X(1, 2)
+```
+
+### Type checking records
+
+Records defined via a `record` statement can be used in type annotations as usual. The unification rules say that a table can unify with a record if the fields
+match, which makes records similar to sealed tables from the type checking perspective. (note, this is hand wavy on subtyping rules)
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+
+The methods defined on the record object are type checked as usual, with one exception - the implicit `self` has the type of the record. This is crucial because
+this is the one big issue we can't resolve with metatable-based OOP for tables, but it works for records because `:` is slightly more magical.
+
+When `self` is explicit, the type needs to be specified manually, e.g. these definitions are equivalent:
+
+```
+function Point:sum(): number
+    return self.x + self.y
+end
+
+function Point.sum(self: Point): number
+    return self.x + self.y
+end
+```
+
+Of course, the type checker also knows that the record type has the metatable with the inferred type of the record shape.
+
+> TODO: Does the type checker need to understand the internal structure of the shape so that type checking works across modules, or is simply modeling this
+> as a metatable sufficient?
+
+### Object modeling
+
+> TODO: records don't support implementation inheritance and why it's a good thing
```

Comment:
Also: We should ensure that you can always use a table type as a way to describe an interface.

Maybe we want some syntax so that it's easy to say that a particular record is expected to be compatible with some other type?

Reply:
Mhm. I was thinking in the past about type-level interfaces. I feel like we should not define this immediately in this proposal but have an obvious route to do that syntactically in the future. Let me think through some options here.

Reply:
Updated, without the specific syntax though - just mentioning we might want this in the future.